frontend/dockerfile/shell/lex_test.go

```diff
@@ -358,6 +358,48 @@ func TestProcessWithMatches(t *testing.T) {
 			},
 			expectedErr: true,
 		},
+		{
+			input:    "${FOO%y*}",
+			envs:     map[string]string{"FOO": "xxyy"},
+			expected: "xxy",
+			matches:  map[string]struct{}{"FOO": {}},
+		},
+		{
+			input:    "${FOO%%y*}",
+			envs:     map[string]string{"FOO": "xxyy"},
+			expected: "xx",
+			matches:  map[string]struct{}{"FOO": {}},
+		},
+		{
+			input:    "${FOO#*x}",
+			envs:     map[string]string{"FOO": "xxyy"},
+			expected: "xyy",
+			matches:  map[string]struct{}{"FOO": {}},
+		},
+		{
+			input:    "${FOO##*x}",
+			envs:     map[string]string{"FOO": "xxyy"},
+			expected: "yy",
+			matches:  map[string]struct{}{"FOO": {}},
+		},
+		{
+			input:    "${FOO#?}",
+			envs:     map[string]string{"FOO": "xxyy"},
+			expected: "xyy",
+			matches:  map[string]struct{}{"FOO": {}},
+		},
+		{
+			input:    "${FOO%x}",
+			envs:     map[string]string{"FOO": "xxyy"},
+			expected: "xxyy",
+			matches:  map[string]struct{}{"FOO": {}},
+		},
+		{
+			input:    "${FOO%%x*}",
+			envs:     map[string]string{"FOO": "xxyy"},
+			expected: "",
+			matches:  map[string]struct{}{"FOO": {}},
+		},
```

Comment:
Can you include some tests for the things like `QuoteMeta`? I noticed that `QuoteMeta` is used, which is likely correct, but these patterns don't have any meta patterns that are special in a regular expression but aren't in a filepath match. Something like the period character is what I'm looking for.

Also, since I think it would be good to split up the regular expression compilation into its own function as shown in the playground link above, I'd also like to see some tests that verify the proper regular expression is generated. I'm not normally the biggest fan of testing internal methods, but I think in this case it would really help with the readability and reliability just because this kind of code can be so tough to verify.

Reply:
I changed the `QuoteMeta` to a list of special characters to escape in the regex builder and also added unit tests for the two helper methods. The tests are small and fast enough to not matter while hopefully handling all edge cases.